File: frontend/src/lib/export.ts
import { parseTokens, SECRET_MASK, type ResolveFn } from './vars'

export type ExportFormat = 'curl' | 'fetch' | 'wget' | 'httpie' | 'powershell'

export interface ExportOptions {
  resolve?: ResolveFn
  keepTokens?: boolean
}

function quote(s: string): string {
  return `'${s.replace(/'/g, `'\\''`)}'`
}

// Like the send path, but a secret never comes out: an export is shared, pasted
// and screenshotted, so it carries the dots.
function render(text: string, opts?: ExportOptions): string {
  if (!opts?.resolve || opts.keepTokens) return text
  const tokens = parseTokens(text)
  if (tokens.length === 0) return text
  let out = ''
  let last = 0
  for (const t of tokens) {
    const r = opts.resolve(t.name)
    out += text.slice(last, t.start)
    out += r ? (r.kind === 'secret' ? SECRET_MASK : r.value) : t.raw
    last = t.end
  }
  return out + text.slice(last)
}

export function exportRequest(
  format: ExportFormat,
  method: string,
  url: string,
  headers: Record<string, string>,
  body: string,
  opts?: ExportOptions
): string {
  url = render(url, opts)
  body = render(body, opts)
  const resolved: Record<string, string> = {}
  for (const [k, v] of Object.entries(headers)) resolved[render(k, opts)] = render(v, opts)
  headers = resolved
  const entries = Object.entries(headers)
  switch (format) {
    case 'curl': {
      const parts = ['curl']
      if (method !== 'GET') parts.push('-X', method)
      for (const [k, v] of entries) parts.push('-H', quote(`${k}: ${v}`))
      if (body) parts.push('--data-raw', quote(body))
      parts.push(quote(url))
      return parts.join(' ')
    }
    case 'fetch': {
      const init: Record<string, unknown> = { method }
      if (entries.length) init.headers = headers
      if (body) init.body = body
      return `fetch(${quote(url)}, ${JSON.stringify(init)})`
    }
    case 'wget': {
      const parts = ['wget']
      if (method !== 'GET') parts.push(`--method=${method}`)
      for (const [k, v] of entries) parts.push(`--header=${quote(`${k}: ${v}`)}`)
      if (body) parts.push(`--body-data=${quote(body)}`)
      parts.push(quote(url))
      return parts.join(' ')
    }
    case 'httpie': {
      const parts = ['http']
      if (body) parts.push(`--raw=${quote(body)}`)
      parts.push(method, quote(url))
      for (const [k, v] of entries) parts.push(quote(`${k}:${v}`))
      return parts.join(' ')
    }
    case 'powershell': {
      const parts = [`Invoke-RestMethod -Method ${method} -Uri ${quote(url)}`]
      if (entries.length) {
        const h = entries.map(([k, v]) => `${k} = ${quote(v)}`).join('; ')
        parts.push(`-Headers @{ ${h} }`)
      }
      if (body) parts.push(`-Body ${quote(body)}`)
      return parts.join(' ')
    }
  }
}

export async function copyToClipboard(text: string): Promise<boolean> {
  try {
    await navigator.clipboard.writeText(text)
    return true
  } catch {
    try {
      const ta = document.createElement('textarea')
      ta.value = text
      ta.style.position = 'fixed'
      ta.style.opacity = '0'
      document.body.appendChild(ta)
      ta.select()
      const ok = document.execCommand('copy')
      document.body.removeChild(ta)
      return ok
    } catch {
      return false
    }
  }
}

File: internal/command/export.go
package command

import (
	"strings"

	"json-inspector/internal/vars"
)

// shellQuote wraps a value in single quotes, closing and reopening them around an embedded quote:
// to a POSIX shell the three pieces are one word and the quote survives as itself.
func shellQuote(s string) string {
	return "'" + strings.ReplaceAll(s, "'", `'\''`) + "'"
}

// psQuote is PowerShell's rule instead: inside single quotes a quote is written twice.
func psQuote(s string) string {
	return "'" + strings.ReplaceAll(s, "'", "''") + "'"
}

// substituteForExport is what an export does with tokens. An export outlives the moment of sending,
// so it goes through the masked substitution: a secret leaves as the mask, never as its value.
func substituteForExport(text string, opts ExportOptions) string {
	if opts.Resolve == nil || opts.KeepTokens {
		return text
	}
	return vars.SubstituteMasked(text, opts.Resolve)
}

// orderedHeaders is the `Record<string, string>` the exporters iterate. Losing a header to a name
// collision would change the command, so it collapses the way a JS object does: a name that is
// already there keeps its position and takes the new value.
type orderedHeaders struct {
	names []string
	vals  map[string]string
}

func (h *orderedHeaders) set(name, value string) {
	if h.vals == nil {
		h.vals = map[string]string{}
	}
	if _, seen := h.vals[name]; !seen {
		h.names = append(h.names, name)
	}
	h.vals[name] = value
}

// entries is Object.entries: the current values, named in JS property order — a name that looks
// like an array index is listed before the rest, whatever position it was set in.
func (h orderedHeaders) entries() []Header {
	out := make([]Header, 0, len(h.names))
	for _, name := range jsPropOrder(h.names) {
		out = append(out, Header{Name: name, Value: h.vals[name]})
	}
	return out
}

// Export renders a request as a command for one tool. It is the inverse of Parse for the cases the
// TS is one for: what a command cannot express — a cookie file, an upload, a flag the parser
// ignored — is simply not written out.
func Export(format Format, req Request, opts ExportOptions) string {
	url := substituteForExport(req.URL, opts)
	body := substituteForExport(req.Body, opts)
	var headers orderedHeaders
	for _, h := range req.Headers {
		headers.set(substituteForExport(h.Name, opts), substituteForExport(h.Value, opts))
	}
	entries := headers.entries()

	switch format {
	case FormatCurl:
		parts := []string{"curl"}
		if req.Method != "GET" {
			parts = append(parts, "-X", req.Method)
		}
		for _, h := range entries {
			parts = append(parts, "-H", shellQuote(h.Name+": "+h.Value))
		}
		if body != "" {
			parts = append(parts, "--data-raw", shellQuote(body))
		}
		parts = append(parts, shellQuote(url))
		return strings.Join(parts, " ")

	case FormatFetch:
		init := jsValue{kind: jsObj}
		init.set("method", jsValue{kind: jsStr, str: req.Method})
		if len(entries) > 0 {
			h := jsValue{kind: jsObj}
			for _, e := range entries {
				h.set(e.Name, jsValue{kind: jsStr, str: e.Value})
			}
			init.set("headers", h)
		}
		if body != "" {
			init.set("body", jsValue{kind: jsStr, str: body})
		}
		return "fetch(" + shellQuote(url) + ", " + jsJSONStringify(init) + ")"

	case FormatWget:
		parts := []string{"wget"}
		if req.Method != "GET" {
			parts = append(parts, "--method="+req.Method)
		}
		for _, h := range entries {
			parts = append(parts, "--header="+shellQuote(h.Name+": "+h.Value))
		}
		if body != "" {
			parts = append(parts, "--body-data="+shellQuote(body))
		}
		parts = append(parts, shellQuote(url))
		return strings.Join(parts, " ")

	case FormatHTTPie:
		parts := []string{"http"}
		if body != "" {
			parts = append(parts, "--raw="+shellQuote(body))
		}
		parts = append(parts, req.Method, shellQuote(url))
		for _, h := range entries {
			parts = append(parts, shellQuote(h.Name+":"+h.Value))
		}
		return strings.Join(parts, " ")

	case FormatPowerShell:
		parts := []string{"Invoke-RestMethod -Method " + req.Method + " -Uri " + psQuote(url)}
		if len(entries) > 0 {
			pairs := make([]string, 0, len(entries))
			for _, h := range entries {
				pairs = append(pairs, psQuote(h.Name)+" = "+psQuote(h.Value))
			}
			parts = append(parts, "-Headers @{ "+strings.Join(pairs, "; ")+" }")
		}
		if body != "" {
			parts = append(parts, "-Body "+psQuote(body))
		}
		return strings.Join(parts, " ")
	}
	return ""
}
